### DISCUSSION TITLE
Graceful shutdown & Forcing exporter to push telemetry

### DISCUSSION THREAD

**Discussion by aignatovich** [open-telemetry/opentelemetry-dotnet#3614] [Q&A] [answered]
**What do I want to do?**
During a shutdown of the application I want to explicitly make sure that logs, metrics, traces are collected and pushed before application is finally disposed.

**What I would like to find out?**
If there an API to explicitly force-push telemetry to the collector?

Usage may look like:

` try
            {
                ...
                app.Run();
            }            
            finally
            {
                Exporter.Dispose();
            }`
            
Can anybody make any suggestions on that?

A workaround that I am not happy about: 
wait for just enough time to make sure that traces are pushed and metrics are collected - then dispose the application.

**Comment by cijothomas** [comment, marked as answer]
There is ForceFlush, but if your goal is to call something at app shutdown, then disposing the provider (MeterProvider, TracerProvider, LoggerFactory) is sufficient.

**Reply by MichaelLogutov** [reply]
How can you do this with prometheus exporter? I want push metrics to specific endpoint during container shutdown.

**Reply by cijothomas** [reply]
You cannot do flush with prometheus exporter (or any exporter that is pull based). 
Prometheus now has native OTLP ingestion, so you can leverage OTLPExporter for Promtheus.
